### PR TITLE
[spaceship][deliver] add app clips support for specifying default experience metadata

### LINKED ISSUE #17639: Deliver error for App containing an App Clip
### New Issue Checklist

- [ ] Updated fastlane to the latest version
- [x] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description

Ran `deliver` this morning and got the following error:

`The request cannot be fulfilled because of the state of another resource. - Submit for review errors found.
The provided entity is missing a required relationship - You must provide a value for the relationship 'appClipVersion' with this request`

It looks like the upload fails due to the app having an `App Clip` target. Is there a way to define the meta data needed for App Clips (screenshot, title, App Clip experiences, domains, ...)?

##### Command executed
```
deliver(
  reset_ratings: false,
  submit_for_review: true,
  automatic_release: false,
  phased_release: true,
  force: true, # Skip HTML report verification
  skip_screenshots: false,
  overwrite_screenshots: true,
  ignore_language_directory_validation: true,
  app_version: version_number,
  build_number: build_number,
  skip_binary_upload: true,
  skip_metadata: false,
  metadata_path: "./fastlane/metadata",
  submission_information: {
    add_id_info_uses_idfa: false
  }
)
```

### Environment

I've redacted the `fastfile` and `appfile` as I don't think it's relevant for this issue.

<details><summary>🚫 fastlane environment 🚫</summary>

### Stack

| Key                         | Value                                                          |
| --------------------------- | -------------------------------------------------------------- |
| OS                          | 11.0.1                                                         |
| Ruby                        | 2.6.3                                                          |
| Bundler?                    | false                                                          |
| Git                         | git version 2.24.3 (Apple Git-128)                             |
| Installation Source         | /usr/local/bin/fastlane                                        |
| Host                        | macOS 11.0.1 (20B29)                                           |
| Ruby Lib Dir                | /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib |
| OpenSSL Version             | LibreSSL 2.8.3                                                 |
| Is contained                | false                                                          |
| Is homebrew                 | false                                                          |
| Is installed via Fabric.app | false                                                          |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/                    |
| Xcode Version               | 12.2                                                           |

### System Locale

| Error                       |
| --------------------------- |
| No Locale with UTF8 found 🚫 |

### fastlane gems

| Gem      | Version | Update-Status      |
| -------- | ------- | ------------------ |
| fastlane | 2.164.0 | 🚫 Update available |

### Loaded fastlane plugins:

| Plugin                  | Version | Update-Status      |
| ----------------------- | ------- | ------------------ |
| fastlane-plugin-badge   | 1.4.0   | ✅ Up-To-Date       |
| fastlane-plugin-bugsnag | 1.4.2   | 🚫 Update available |

<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| did_you_mean              | 1.3.0        |
| slack-notifier            | 2.3.2        |
| rouge                     | 2.0.7        |
| xcpretty                  | 0.3.0        |
| terminal-notifier         | 2.0.0        |
| terminal-table            | 1.8.0        |
| plist                     | 3.5.0        |
| addressable               | 2.7.0        |
| multipart-post            | 2.0.0        |
| word_wrap                 | 1.0.0        |
| colored                   | 1.2          |
| highline                  | 1.7.10       |
| commander-fastlane        | 4.4.6        |
| faraday_middleware        | 1.0.0        |
| gh_inspector              | 1.1.3        |
| rubyzip                   | 2.3.0        |
| security                  | 0.1.3        |
| xcpretty-travis-formatter | 1.0.0        |
| uber                      | 0.1.0        |
| declarative-option        | 0.1.0        |
| representable             | 3.0.4        |
| retriable                 | 3.1.2        |
| mini_mime                 | 1.0.2        |
| httpclient                | 2.8.3        |
| google-api-client         | 0.38.0       |
| nanaimo                   | 0.3.0        |
| colored2                  | 3.1.2        |
| claide                    | 1.0.3        |
| CFPropertyList            | 3.0.2        |
| atomos                    | 0.1.3        |
| xcodeproj                 | 1.19.0       |
| unicode-display_width     | 1.7.0        |
| public_suffix             | 4.0.6        |
| tty-screen                | 0.8.1        |
| tty-cursor                | 0.7.1        |
| tty-spinner               | 0.9.3        |
| babosa                    | 1.0.4        |
| excon                     | 0.78.0       |
| unf_ext                   | 0.0.7.7      |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20190701 |
| http-cookie               | 1.0.3        |
| ruby2_keywords            | 0.0.2        |
| faraday                   | 1.1.0        |
| faraday-cookie_jar        | 0.0.7        |
| fastimage                 | 2.2.0        |
| json                      | 2.3.1        |
| mini_magick               | 4.10.1       |
| dotenv                    | 2.7.6        |
| bundler                   | 1.17.3       |
| naturally                 | 2.2.0        |
| simctl                    | 1.6.8        |
| jwt                       | 2.2.2        |
| declarative               | 0.0.20       |
| multi_json                | 1.15.0       |
| signet                    | 0.14.0       |
| os                        | 1.1.1        |
| memoist                   | 0.16.2       |
| googleauth                | 0.14.0       |
| rake                      | 13.0.1       |
| digest-crc                | 0.6.1        |
| google-cloud-errors       | 1.0.1        |
| google-cloud-env          | 1.4.0        |
| google-cloud-core         | 1.5.0        |
| google-cloud-storage      | 1.29.1       |
| emoji_regex               | 3.0.0        |
| aws-eventstream           | 1.1.0        |
| aws-sigv4                 | 1.2.2        |
| aws-partitions            | 1.384.0      |
| jmespath                  | 1.4.0        |
| aws-sdk-core              | 3.109.1      |
| aws-sdk-kms               | 1.39.0       |
| aws-sdk-s3                | 1.83.1       |
| forwardable               | 1.2.0        |
| logger                    | 1.3.0        |
| stringio                  | 0.0.2        |
| ipaddr                    | 1.2.2        |
| openssl                   | 2.1.2        |
| ostruct                   | 0.1.0        |
| strscan                   | 1.0.0        |
| date                      | 2.0.0        |
| fileutils                 | 1.1.0        |
| etc                       | 1.0.1        |
| io-console                | 0.4.7        |
| zlib                      | 1.0.0        |
| libxml-ruby               | 3.1.0        |
| rexml                     | 3.1.9        |
| psych                     | 3.1.0        |
| mutex_m                   | 0.1.0        |
| webrick                   | 1.4.2        |
| curb                      | 0.9.10       |
| badge                     | 0.12.0       |
| fastlane-plugin-badge     | 1.4.0        |
| xml-simple                | 1.1.5        |
| fastlane-plugin-bugsnag   | 1.4.2        |
</details>

*generated on:* **2020-11-20**
</details>

### PR DESCRIPTION
**Pull request by blesson3**
### Checklist

- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

There is currently no app clips functionality within fastlane. My team and I ran into the issue described in #17639 and had to do manual releases (not fun). I decided to tackle this to help lessen the barrier a bit for others utilizing app clips.

There existed another seemingly abandoned WIP PR #19043 that initially implemented app clips using the web session and private APIs. The ASC API then came out with full app clips support, but with changed models/api-endpoints. This in turn required that PR to re-implement much of it's content, which has not yet been done.

This PR looks to implement the goals of #19043, but with additional `deliver` support so developers can immediately start working with app clips metadata automation via simple `Fastfile` updates.

Resolves #17639

### Description

This PR is one of a few app clips related changes I have made. I am presenting the PRs in smaller pieces so the implications of each PR can be discussed in a tighter scope.

The following app clip default experience functionality has been added to `deliver`:
- option to specify action (PLAY, VIEW, ...)
- option to specify localized subtitle
- option to specify localized default experience header image

If none of these options are specified, then fastlane will copy over the app clip default experience metadata from the live version. This is similar to the default behavior of the ASC UI discussed [here](https://github.com/fastlane/fastlane/issues/17639#issuecomment-1130196397).

Example:
```rb
deliver(
  app_clip_default_experience_action: "PLAY",
  app_clip_default_experience_subtitle: {
    "en-US": "this is a test from fastlane",
    "es-MX": "está es una prueba de fastlane"
  },
  app_clip_header_images_path: "./fastlane/app-clip-header-images"
)
```

The localized header images are read from the filesystem same as screenshots:

```sh
$ tree app-clip-header-images/
app-clip-header-images
├── en-US
│   └── random-header-image.jpg
└── es-MX
    └── another-header-image.jpg

2 directories, 2 files
```

These options relate to the ASC UI in the "App Clips" section under a specific App Store version: CleanShot 2022-05-19 at 12 12 29@2x

### Discussion

Prior discussion can be [found here](https://github.com/fastlane/fastlane/issues/17639#issuecomment-1121590421).

This PR does NOT implement anything wrt app clip **advanced experiences**. This PR focuses on the default experience associated with the App Store versions only. The advanced experiences are a bit of different beast. If there's interest and it makes sense, I may consider adding that in another PR since I am now familiar with the ASC APIs.

The following should be addressed before this PR is ready to be merged:

1. ~~Apple allows more than one app clip target per app, this PR uses the [first listed app clip only](https://github.com/qatonda/fastlane/blob/9bf2f6bd8bf8dd2bef97dca0e11c6a44928c79a5/deliver/lib/deliver/upload_app_clip_default_experience_metadata.rb#L26).~~
  - Resolved in [this comment](https://github.com/fastlane/fastlane/pull/20300#issuecomment-1133162510)

<a href="#discussion-point-2" id="discussion-point-2">#</a> 2.  ~~docs: I see [generated documentation here](https://github.com/fastlane/docs/tree/master/docs/generated), but it's unclear how I update them. I would be happy to update the docs and add an app clips section if someone pointed me in the right direction.~~
  - Resolved with [a commit to update docs](https://github.com/fastlane/fastlane/pull/20300/commits/de5f6d90cc72a7506bc19120cee64a21d4a7c4d4)

3. Do we want the app clips metadata to show up in the interactive HTML metadata report?
    - I have not ever used this feature, but I noticed it while testing.

#### Code Quality

> Disclaimer: I am not a ruby programmer, so I will have missed a lot of ruby-isms in this PR, but I tried my best to keep the code style consistent with the rest of the codebase.

I have not yet added tests for these changes. I have tested them using a small test project and my own projects in production.

Some of the code in this PR has been copied from other relevant parts of the codebase.

##### For example:

The new file [`spaceship/connect_api/models/app_clip_header_image.rb`](https://github.com/fastlane/fastlane/blob/55fb13119ff9b65c865977be6390ecd6cd738360/spaceship/lib/spaceship/connect_api/models/app_clip_header_image.rb) has code copied from [`.../models/app_screenshot.rb`](https://github.com/fastlane/fastlane/blob/master/spaceship/lib/spaceship/connect_api/models/app_screenshot.rb) since the upload functionality was the same. I decided not to inherit or subclass because I was sure that making the upload process more general would be more work than it was worth. I am absolutely open to hearing counter arguments.

The PR #17639 has a [similar comment about extracting common functionality here](https://github.com/fastlane/fastlane/pull/19043#pullrequestreview-698633451).

#### Basic Testing Setup

1. Update `Gemfile` to 👇 and run `bundle install`, `bundle update fastlane`, or `bundle update`
```rb
gem "fastlane", :git => 'https://github.com/qatonda/fastlane.git', :ref => "matt/app-clips-default-experience"
```
2. Upload a binary to ASC with an app clip target
    - you can follow [a project tutorial provided by Apple here](https://developer.apple.com/documentation/swiftui/fruta_building_a_feature-rich_app_with_swiftui)

#### What to Test

Before specifying any `deliver` options:
- Confirm that fastlane uses the live version's app clip metadata for the current edit version
    - It should copy over the action, subtitle, and header image of the live version
    - This is not already the default when the AS version is created via the ASC APIs, thus the following issue #17639

After specifying the app clip default experience `deliver` options:
- Confirm the app clip default experience metadata specified in the `deliver` options reflects the ASC UI after running `deliver`
- Confirm that fastlane enforces that the app clip header image is exactly 1800x1200 in size
- Confirm that fastlane enforces only one app clip header image per locale
- Confirm that the additional fastlane messages via `UI.{message,important,...}` are appropriate

### CODE REVIEW

File: deliver/lib/deliver/upload_app_clip_default_experience_metadata.rb [hunk 1]
```diff
@@ -102,5 +120,49 @@ def upload_subtitle_localizations(app_clip_default_experience:, subtitle_localiz
...
+        # default_experience.update(attributes: { action: live_version_default_experience.action })
+        # UI.message("Added app clip default experience action from live version #{live_version.version}")
+
+        # # copy the live version localizations
+        # live_version_localizations = Spaceship::ConnectAPI::AppClipDefaultExperience.get(app_clip_default_experience_id: live_version_default_experience.id, includes: 'appClipDefaultExperienceLocalizations').app_clip_default_experience_localizations
+
+        # live_version_localizations.each do |live_localization|
+        #   attributes = { subtitle: live_localization.subtitle, locale: live_localization.locale }
+        #   Spaceship::ConnectAPI::AppClipDefaultExperienceLocalizations.create(default_experience_id: default_experience.id, attributes: attributes)
+        # end
```
Comment (blesson3) [inline deliver/lib/deliver/upload_app_clip_default_experience_metadata.rb:161]: I'm planning to remove this commented code once I can confirm the default experience template option works as I expect.

**Comment by EmDee** [comment]
@blesson3 I've ran into an error.

I'm using the latest commit (`028092faaede7d3cc4222618ffa5bcfdb10d4e60`): `Using fastlane 2.206.0 from https://github.com/qatonda/fastlane.git (at matt/app-clips-default-experience@028092f)`

```
bundler: failed to load command: fastlane (/Users/pace/.rbenv/versions/2.6.4/lib/ruby/gems/2.6.0/bin/fastlane)
/Users/pace/.rbenv/versions/2.6.4/lib/ruby/gems/2.6.0/bundler/gems/fastlane-028092faaede/deliver/lib/deliver/upload_app_clip_default_experience_metadata.rb:96:in `upload_subtitle_localizations': [!] undefined method `id' for nil:NilClass (NoMethodError)
```

**Comment by blesson3** [comment]
> @blesson3 I've ran into an error.
> 
> I'm using the latest commit (`028092faaede7d3cc4222618ffa5bcfdb10d4e60`): `Using fastlane 2.206.0 from https://github.com/qatonda/fastlane.git (at matt/app-clips-default-experience@028092f)`
> 
> ```
> bundler: failed to load command: fastlane (/Users/pace/.rbenv/versions/2.6.4/lib/ruby/gems/2.6.0/bin/fastlane)
> /Users/pace/.rbenv/versions/2.6.4/lib/ruby/gems/2.6.0/bundler/gems/fastlane-028092faaede/deliver/lib/deliver/upload_app_clip_default_experience_metadata.rb:96:in `upload_subtitle_localizations': [!] undefined method `id' for nil:NilClass (NoMethodError)
> ```

@EmDee Thanks for the feedback. I've fixed the issue in 15506706d975e5da7ee095b4af057c2d925907c1. I'm testing the changes now on my own projects to confirm the issue is fixed.

**Comment by blesson3** [comment]
### Re: apps having more than one app clip target

_(no action needed here on this PR, but this addresses one of the points from the PR description)_

It seems that building an app with more than one app clip target produces the following error from xcode:
```
❌  error: App contains multiple App Clips. It can only have one.
```

It seems impossible to upload a binary with more than one app clip target. What's confusing about this is that the ASC API [allows for more than one app clip to be returned](https://developer.apple.com/documentation/appstoreconnectapi/search_an_app_s_app_clips) for any given app. We handle this by [picking only the first app clip](https://github.com/qatonda/fastlane/blob/9bf2f6bd8bf8dd2bef97dca0e11c6a44928c79a5/deliver/lib/deliver/upload_app_clip_default_experience_metadata.rb#L26) from the response. If this changes in the future, we can address it then.

**Comment by EmDee** [comment]
> @EmDee Thanks for the feedback. I've fixed the issue in [1550670](https://github.com/fastlane/fastlane/commit/15506706d975e5da7ee095b4af057c2d925907c1) and [fc088f0](https://github.com/fastlane/fastlane/commit/fc088f0e2362fdb2749c8184bd31ad5e946f3c09). I've tested the changes using my own projects and it seems to work well. Please test and lmk!

Thank you, will do! Probably around Tuesday.

**Comment by EmDee** [comment]
@blesson3 Looks like you did another rebase?

I was still on the version of 2-3 weeks ago. I just ran into a different error message, not sure, if you've already fixed this one:

```
/Users/user/.gem/ruby/2.6.0/bundler/gems/fastlane-fc088f0e2362/spaceship/lib/spaceship/connect_api/tunes/tunes.rb:1291:in `post_app_clip_default_experience': undefined method `data' for #<Hash:0x00007f7d1b86cd20> (NoMethodError)
	from /Users/user/.rbenv/versions/2.6.4/lib/ruby/2.6.0/forwardable.rb:230:in `post_app_clip_default_experience'
	from /Users/user/.gem/ruby/2.6.0/bundler/gems/fastlane-fc088f0e2362/spaceship/lib/spaceship/connect_api/models/app_clip_default_experience.rb:25:in `create'
	from /Users/user/.gem/ruby/2.6.0/bundler/gems/fastlane-fc088f0e2362/deliver/lib/deliver/upload_app_clip_default_experience_metadata.rb:51:in `upload_metadata'
```

**Comment by ConfusedVorlon** [comment]
This is great. I just hit the relationship error which stopped me submitting a release for an app with an appclip.
I switched to your branch, made no other changes and everything worked great.
(the app used the details from my previous submission)

Brilliant - and Thank You :)

I hope this gets merged soon!

**Comment by EmDee** [comment]
@blesson3 Great, I'll give the update a test run this Friday.

**Comment by EmDee** [comment]
@blesson3 Did two releases without a hitch!

**Comment by EmDee** [comment]
We've had couple releases now, which all worked great.

@blesson3 Any open tasks on your end or can this be cleaned up and sent to review?

**Comment by blesson3** [comment]
Great to hear. The only thing that might be holding up review is formal documentation. Since app clips is a new class of functionality here, it may be difficult for people to know how to use these new features. Although I'm down to push that into a separate PR so this one can be merged more quickly.

Per [discussion point #2](https://github.com/fastlane/fastlane/pull/20300#discussion-point-2), I was unable to find where I should add documentation. Ultimately I think it should go on [docs.fastlane.tools/actions/deliver](https://docs.fastlane.tools/actions/deliver/).

Pinging maintainers for clarification: @joshdholtz @getaaron

**Comment by EmDee** [comment]
Small bump for these great changes and pinging the core team in the hopes that someone can clarify the documentation issue mentioned in https://github.com/fastlane/fastlane/pull/20300#issuecomment-1171764634: @joshdholtz @getaaron @rogerluan @milch

**Comment by rogerluan** [comment]
@blesson3 Answering specifically the documentation concerns, the regular parameters/options/descriptions documentation will be autogenerated from the lane metadata. However, I agree with you that extra documentation would be super helpful here, and I'd also expect it to live in the link you mentioned, in which case you'd want to edit `fastlane/lib/fastlane/actions/docs/upload_to_app_store.md.erb` in this repository 😊 it will then be autogenerated in https://github.com/fastlane/docs when creating the next release :)

Hope this helps!

**Comment by landsman** [comment]
We are now investigating possibilities how to release App Clips with Fastline and this is probably what we are looking for 🙏

**Comment by EmDee** [comment]
Still using this branch without any bigger hitch. 

@blesson3 Would you mind rebasing your branch onto the fastlane's master? Some nice changes have been added, e.g. the use of `altool`.

**Comment by blesson3** [comment]
> @blesson3 Would you mind rebasing your branch onto the fastlane's master? Some nice changes have been added, e.g. the use of altool.

@EmDee Done! Been busy so I haven't yet had the time to invest into writing the documentation.

**Comment by EmDee** [comment]
Thanks so much @blesson3!

I can only state again that this MR is pure 🔥

The changes here have saved us so much time in the past couple of months! 🚀

**Comment by EmDee** [comment]
@blesson3 Just a quick follow-up question:

We started to add a bunch of languages to our app. As far as I can tell from your changes the `app_clip_default_experience_subtitle` must be set via the `deliver` command, correct? There is no way to have it automatically be populated via some metadata folder structure? (Would definitely be a nice additional follow-up in the future)

**Comment by blesson3** [comment]
> @blesson3 Just a quick follow-up question:
> 
> We started to add a bunch of languages to our app. As far as I can tell from your changes the `app_clip_default_experience_subtitle` must be set via the `deliver` command, correct? There is no way to have it automatically be populated via some metadata folder structure? (Would definitely be a nice additional follow-up in the future)

@EmDee That's correct in the PR's current implementation. I can definitely explore in the future populating app clip default experience metadata from a folder structure.

**Comment by EmDee** [comment]
@blesson3 Since we are still dependent on these changes, do you mind doing another rebase (as their are some critical fixes from [2.212.1](https://github.com/fastlane/fastlane/releases/tag/2.212.1) we also need.

**Comment by EmDee** [comment]
Also maybe we can have this PR looked at, @joshdholtz? Or possibly someone else.

**Comment by blesson3** [comment]
@EmDee Sorry for the delayed response, I've recently moved jobs outside of iOS development and I'm no longer with the company that hosts this PR's source code. I just rebased with `upstream/master` so it should be up to date again.

Feel free to create a new PR using the commits in this one if you want to own this change. I'm not able to do further work or changes based on maintainer feedback (whenever that might be), but I can continue to rebase.

Note that there are several other changes not contained within this PR, like adding an [option to specify beta app clip invocations](https://github.com/qatonda/fastlane/commit/8ac09da3bdc5199b8e7bee0d2976acc86e09f41c) and for [app store review information](https://github.com/qatonda/fastlane/commit/48802b5d5ad56151ea1b72e53f936786ca6bb09a), which can be found in the branch [`qatonda/fastlane:matt/app-clips`](https://github.com/qatonda/fastlane/commits/matt/app-clips). I was planning to open separate PRs after this one was merged. Since I'm not able to see those changes through to master, someone else will have to take over these changes.

**Comment by EmDee** [comment]
@blesson3 A great colleague of mine, @telip007, has worked on top of your changes here and fixed all known issues to us. We've opened up a new PR (https://github.com/fastlane/fastlane/pull/21180) where hopefully someone from Fastlane's team will take a look at the changes.

If you want, we can close up this PR here then.

**Pull request closed without merging**